martin/src/cog/mod.rs

```diff
@@ -0,0 +1,383 @@
+        let ifd = self.meta.zoom_and_ifd.get(&(xyz.z)).ok_or_else(|| {
+            CogError::ZoomOutOfRange(
+                xyz.z,
+                self.path.clone(),
+                self.meta.min_zoom,
+                self.meta.max_zoom,
+            )
+        })?;
+
+        decoder
+            .seek_to_image(*ifd)
+            .map_err(|e| CogError::IfdSeekFailed(e, *ifd, self.path.clone()))?;
+
+        let tiles_across = self
+            .meta
+            .zoom_and_tile_across_down
+            .get(&(xyz.z))
+            .ok_or_else(|| {
+                CogError::ZoomOutOfRange(
+                    xyz.z,
+                    self.path.clone(),
+                    self.meta.min_zoom,
+                    self.meta.max_zoom,
+                )
+            })?
+            .0;
+        let tile_idx = xyz.y * tiles_across + xyz.x;
+        let decode_result = decoder
+            .read_chunk(tile_idx)
+            .map_err(|e| CogError::ReadChunkFailed(e, tile_idx, *ifd, self.path.clone()))?;
+        let color_type = decoder
+            .colortype()
+            .map_err(|e| CogError::InvalidTifFile(e, self.path.clone()))?;
+
+        let tile_width = decoder.chunk_dimensions().0;
+        let tile_height = decoder.chunk_dimensions().1;
+        let (data_width, data_height) = decoder.chunk_data_dimensions(tile_idx);
+
+        //do more research on the not u8 case, is this the right way to do it?
+        let png_file_bytes = match (decode_result, color_type) {
+            (DecodingResult::U8(vec), tiff::ColorType::RGB(_)) => rgb_to_png(
+                vec,
+                (tile_width, tile_height),
+                (data_width, data_height),
+                3,
+                self.meta.nodata.map(|v| v as u8),
+                &self.path,
+            ),
+            (DecodingResult::U8(vec), tiff::ColorType::RGBA(_)) => rgb_to_png(
+                vec,
+                (tile_width, tile_height),
+                (data_width, data_height),
+                4,
+                self.meta.nodata.map(|v| v as u8),
+                &self.path,
+            ),
+            (_, _) => Err(CogError::NotSupportedColorTypeAndBitDepth(
+                color_type,
+                self.path.clone(),
+            )),
+            // do others in next PRs, a lot of disscussion would be needed
+        }?;
+        Ok(png_file_bytes)
+    }
+}
+
+fn rgb_to_png(
+    vec: Vec<u8>,
+    result_dims: (u32, u32),
+    chunk_dims: (u32, u32),
+    chunk_components_count: u32,
+    nodata: Option<u8>,
+    path: &Path,
+) -> Result<Vec<u8>, CogError> {
+    let (data_width, data_height) = chunk_dims;
+    let (tile_width, tile_height) = result_dims;
+    let is_padded = data_width != tile_width;
+    let need_add_alpha = chunk_components_count != 4;
+    let default_value = 0;
+
+    let pixels = if nodata.is_some() || need_add_alpha || is_padded {
+        let mut result_vec = vec![default_value; (tile_width * tile_height * 4) as usize];
```

Comment:
nit: imo inlinining this is more readable. What do you think?

```suggestion
    let pixels = if nodata.is_some() || need_add_alpha || is_padded {
        let mut result_vec = vec![0; (tile_width * tile_height * 4) as usize];
```

Reply:
Both Ok I guess.